File: daily/cpp/1609_even_odd_tree.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left, *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  bool isEvenOddTree(TreeNode *root) {
    queue<TreeNode *> q;
    int level = 0;
    q.push(root);

    while (!q.empty()) {
      int size = q.size();
      int prev = (level % 2 == 0) ? numeric_limits<int>::min()
                                  : numeric_limits<int>::max();

      for (int i = 0; i < size; ++i) {
        TreeNode *node = q.front();
        q.pop();

        if ((level % 2 == 0 && (node->val % 2 == 0 || node->val <= prev)) ||
            (level % 2 == 1 && (node->val % 2 == 1 || node->val >= prev))) {
          return false;
        }

        prev = node->val;

        if (node->left) {
          q.push(node->left);
        }
        if (node->right) {
          q.push(node->right);
        }
      }
      ++level;
    }

    return true;
  }
};

File: daily/rust/1609_even_odd_tree.rs
use std::{cell::RefCell, rc::Rc};

#[derive(Debug, PartialEq, Eq)]
struct TreeNode {
    val: i32,
    left: Option<Rc<RefCell<TreeNode>>>,
    right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new() {
        TreeNode {
            val: 0,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn is_even_odd_tree(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let mut queue = vec![root];
        let mut level = 0;
        while !queue.is_empty() {
            let mut prev = if level % 2 == 0 { i32::MIN } else { i32::MAX };
            let mut next = vec![];
            for node in queue {
                if let Some(node) = node {
                    let node = node.borrow();
                    if level % 2 == 0 {
                        if node.val % 2 == 0 || node.val <= prev {
                            return false;
                        }
                    } else {
                        if node.val % 2 == 1 || node.val >= prev {
                            return false;
                        }
                    }
                    prev = node.val;
                    if let Some(left) = &node.left {
                        next.push(Some(left.clone()));
                    }
                    if let Some(right) = &node.right {
                        next.push(Some(right.clone()));
                    }
                }
            }
            queue = next;
            level += 1;
        }
        true
    }
}
